src/DGmethods/DGmodel.jl

```diff
@@ -0,0 +1,250 @@
+struct DGModel{BL,G,NF,GNF}
+  balancelaw::BL
+  grid::G
+  divnumflux::NF
+  gradnumflux::GNF
+end
+
+function (dg::DGModel)(dQdt, Q, param, t; increment=false)
+  bl = dg.balancelaw
+  device = typeof(Q.Q) <: Array ? CPU() : CUDA()
+
+  grid = dg.grid
+  topology = grid.topology
+
+  dim = dimensionality(grid)
+  N = polynomialorder(grid)
+  Nq = N + 1
+  Nqk = dim == 2 ? 1 : Nq
+  Nfp = Nq * Nqk
+  nrealelem = length(topology.realelems)
+
+  Qvisc = param.diff
+  auxstate = param.aux
+
+  DFloat = eltype(Q)
+  nviscstate = num_diffusive(bl, DFloat)
+
+  lgl_weights_vec = grid.ω
+  Dmat = grid.D
+  vgeo = grid.vgeo
+  sgeo = grid.sgeo
+  vmapM = grid.vmapM
+  vmapP = grid.vmapP
+  elemtobndy = grid.elemtobndy
+  polyorder = polynomialorder(dg.grid)
+
+  ### update aux variables
+  if hasmethod(update_aux!, Tuple{typeof(bl), Vars, Vars, Vars, DFloat})
+    @launch(device, threads=(Np,), blocks=nrealelem,
+      knl_apply_aux!(bl, Val(dim), Val(polyorder), update_aux!, Q.Q, Qvisc.Q, auxstate.Q, t, topology.realelems)) 
+  end
+
+  ########################
+  # Gradient Computation #
+  ########################
+  MPIStateArrays.start_ghost_exchange!(Q)
+
+  if nviscstate > 0
+
+    @launch(device, threads=(Nq, Nq, Nqk), blocks=nrealelem,
+            volumeviscterms!(bl, Val(dim), Val(polyorder), Q.Q, Qvisc.Q, auxstate.Q, vgeo, t, Dmat,
+                             topology.realelems))
+
+    MPIStateArrays.finish_ghost_recv!(Q)
+
+    @launch(device, threads=Nfp, blocks=nrealelem,
+            faceviscterms!(bl, Val(dim), Val(polyorder), dg.gradnumflux, 
+                           Q.Q, Qvisc.Q, auxstate.Q,
+                           vgeo, sgeo, t, vmapM, vmapP, elemtobndy,
+                           topology.realelems))
+
+    MPIStateArrays.start_ghost_exchange!(Qvisc)
+  end
+
+  ###################
+  # RHS Computation #
+  ###################
+  @launch(device, threads=(Nq, Nq, Nqk), blocks=nrealelem,
+          volumerhs!(bl, Val(dim), Val(polyorder), dQdt.Q, Q.Q, Qvisc.Q, auxstate.Q,
```

Comment:
I still would prefer something like `volume_tendency` or something similar throughout. (It'll avoid a lot of confusion with users.)